tic_tac_toe/game.py

```diff
@@ -63,42 +68,107 @@ def _check_winning_combinations(board, player):
+        ((2, 0), (2, 1), (2, 2)),
+        ((0, 0), (1, 0), (2, 0)), # verticals
+        ((0, 1), (1, 1), (2, 1)),
+        ((0, 2), (1, 2), (2, 2)),
+        ((0, 0), (1, 1), (2, 2)), # diagonals
+        ((0, 2), (1, 1), (2, 0))
+    ]
+
+    for combination in combinations:
+        if _is_winning_combination(board, combination, player):
+            return player
+    return None
 
 # public interface
 def start_new_game(player1, player2):
     """
     Creates and returns a new game configuration.
     """
-    pass
-
+    board = [['-']*3 for _ in range(3)]
+    game = {
+        'player1': player1,
+        'player2': player2,
+        'board': board,
+        'next_turn': player1,
+        'winner': None
+    }
+    
+    return game
 
 def get_winner(game):
     """
     Returns the winner player if any, or None otherwise.
     """
-    pass
-
+    for player in [game['player1'], game['player2']]:
+        if _check_winning_combinations(game['board'], player):
+            return player
+    return None
 
 def move(game, player, position):
     """
     Performs a player movement in the game. Must ensure all the pre requisites
     checks before the actual movement is done.
     After registering the movement it must check if the game is over.
     """
-    pass
-
+    # check if game is already over
+    if game['winner'] or _board_is_full(game['board']):
+        raise InvalidMovement('Game is over.')
+    
+    # check if an illegal movement is being attempted
+    if not _position_is_valid(position):
+        raise InvalidMovement('Position out of range.')
+    if not _position_is_empty_in_board(position, game['board']):
+        raise InvalidMovement('Position already taken.')
+    if game['next_turn'] != player:
+        raise InvalidMovement('"{}" moves next.'.format(game['next_turn']))
+    
+    # the move is valid so add either an 'X' or 'O' to the game board
+    game['board'][position[0]][position[1]] = player
+    
+    # see if it was a winning move, or if it is a tie (board is full)
+    winner = get_winner(game)
+    if winner:
+        game['winner'] = winner
+        raise GameOver('"{}" wins!'.format(winner))
+    if _board_is_full(game['board']):
+        raise GameOver('Game is tied!')
+        
+    # give the next player a turn
+    swap_players = {'X': 'O', 'O': 'X'}
+    game['next_turn'] = swap_players[player]
+    return 
 
 def get_board_as_string(game):
     """
     Returns a string representation of the game board in the current state.
     """
-    pass
-
+    entry_sep = '  |  '
+    row_sep = '--------------'
+    rows = []
+    
+    # format each row of the game board
+    for row in game['board']:
+        row_string = entry_sep.join(row)
+        rows.append(row_string)
+    
+    # add row separators and newlines for pretty board printing
+    rows.insert(3, '')
+    rows.insert(2, row_sep)
+    rows.insert(1, row_sep)
+    rows.insert(0, '')
+    return os.linesep.join(rows)
```

Comment:
Interesting way to return the string version of the board. I think laying out the board in a multi-line string and using .format to input each value is more readable, but nice to see that this method works as well! :+1: